src/Apropos/Gen.hs

```diff
@@ -135,40 +202,45 @@ retryLimit lim g done = go lim
     then pure ()
     else failWithFootnote $ "expected: " <> show l <> " === " <> show r
 
-genProp :: Gen a -> Property
-genProp g = H.property $ void $ runExceptT $ gen g
-
 data GenException = GenException String | Retry deriving stock (Show)
 
-type GenContext = PropertyT IO
+type GenContext = H.Gen
 
-type Generator = ExceptT GenException GenContext
+type Generator = ExceptT GenException (WriterT (Set String) GenContext)
 
 gen :: Gen a -> Generator a
-gen (Free (Label s next)) = H.label (fromString s) >> gen next
-gen (Free (FailWithFootnote s _)) = H.footnote s >> H.failure
-gen (Free (GenBool next)) = lift (H.forAll HGen.bool) >>= (gen . next)
+gen (Free (Label s next)) = lift (tell (Set.singleton s)) >> gen next
+gen (Free (FailWithFootnote s _)) = throwE $ GenException s
+gen (Free (GenBool next)) = lift HGen.bool >>= (gen . next)
 gen (Free (GenInt r next)) = do
-  i <- lift (H.forAll $ HGen.int (hRange r))
+  i <- lift $ HGen.int (hRange r)
   gen $ next i
 gen (Free (GenList r g next)) = do
   let gs = int r >>= \l -> replicateM l g
-  gen gs >>>= next
+  gen gs >=>= next
 gen (Free (GenShuffle ls next)) = do
-  s <- lift $ H.forAll $ HGen.shuffle ls
+  s <- lift $ HGen.shuffle ls
   gen $ next s
 gen (Free (GenElement ls next)) = do
-  s <- lift $ H.forAll $ HGen.element ls
+  s <- lift $ HGen.element ls
   gen $ next s
 gen (Free (GenChoice gs next)) = do
   let l = length gs
   if l == 0
     then throwE $ GenException "GenChoice: list length zero"
     else do
-      i <- lift (H.forAll $ HGen.int (HRange.linear 0 (l -1)))
+      i <- lift (HGen.int (HRange.linear 0 (l -1)))
       (gs !! i) >>== next
 gen (Free (GenFilter c g next)) = do
-  genFilter' c g >>>= next
+  let liftC :: forall a. (a -> Bool) -> (Either GenException a, Set String) -> Bool
+      liftC f (Right a, _) = f a
+      liftC _ _ = False
+  -- This stops the doom shrinking but we really want HGen.filter to throw a retry when it fails
+  -- instead its failure mode fails the whole generator when we may be in a recoverable position if we were to retry.
+  (ee,labels) <- lift $ lift $ HGen.filter (liftC c) $  runWriterT (runExceptT $ gen g)
+  case ee of
+    Left e -> throwE e
+    Right a -> lift (tell labels) >> gen (next a)
```

Comment:
This is nice! Using `HGen.filter` gets us nicer filtering behaviour - except... `HGen.filter` causes us to fail completely if we run out of retries. What we want is to catch the `filter` failing and throw a `Retry` rather than stopping completely.